orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -0,0 +1,255 @@
+        input_args.push_back(outputs[0]);
+      }
+    }
+  };
+
+  std::vector<NodeArg*> input_args;
+  AddInputEvent(graph, "RecordEvent", false /* is_forward */, input_args, new_input_names);
+
+  if (send_bw) {
+    // if we have send op in backward pass (at the end of the graph), add the RecordEvent op after that.
+    input_args.insert(std::end(input_args),
+                      std::begin(send_bw->MutableOutputDefs()),
+                      std::end(send_bw->MutableOutputDefs()));
+  }
+  add_leaf_node(input_args);
+  graph.AddNode(graph.GenerateNodeName("RecordEvent"),
+                "RecordEvent",
+                "Backward pass",
+                input_args,
+                {},
+                nullptr,
+                kMSDomain);
+}
+
+Status AddWaitForward(Graph& graph, Node* recv_fw, std::vector<std::string>& new_input_names) {
+  auto update_wait_input_output = [&](NodeArg* old_input,
+                                      std::vector<NodeArg*>& input_args,
+                                      std::vector<NodeArg*>& output_args) -> NodeArg& {
+    const auto& new_name = graph.GenerateNodeArgName(old_input->Name());
+    ONNX_NAMESPACE::TypeProto type_proto(*(old_input->TypeAsProto()));
+
+    auto& wait_output = graph.GetOrCreateNodeArg(new_name, &type_proto);
+    output_args.push_back(&wait_output);
+    input_args.push_back(old_input);
+    return wait_output;
+  };
+
+  if (recv_fw) {
+    // if we have recv op in forward pass (at the begining of the graph), add the WaitEvent op before that.
+    std::vector<NodeArg*> input_args;
+    std::vector<NodeArg*> output_args;
+    AddInputEvent(graph, "WaitEvent", true /* is_forward */, input_args, new_input_names);
+
+    // recv's first input is the signal input. Re-direct it to WaitEvent's input.
+    auto& input_signal = recv_fw->MutableInputDefs()[0];
+    auto& wait_output = update_wait_input_output(input_signal, input_args, output_args);
+    input_signal = &wait_output;
+
+    graph.AddNode(graph.GenerateNodeName("WaitEvent"),
+                  "WaitEvent",
+                  "",
+                  input_args,
+                  output_args,
+                  nullptr,
+                  kMSDomain);
+  } else {
+    // the first stage doesn't have recv_fw. Add Wait for all inputs.
+    std::vector<NodeArg*> input_args;
+    std::vector<NodeArg*> output_args;
+    AddInputEvent(graph, "WaitEvent", true /* is_forward */, input_args, new_input_names);
+    const std::vector<const NodeArg*>& graph_inputs = graph.GetInputs();
+
+    ORT_RETURN_IF_NOT(
+      graph_inputs.size() > 0, "Graph ", graph.Name(), " doesn't have any inputs.");
+
+    for (auto& input_arg : graph_inputs) {
+      NodeArg* mutable_input = graph.GetNodeArg(input_arg->Name());
+      auto& wait_output = update_wait_input_output(mutable_input, input_args, output_args);
+      std::vector<Node*> nodes = graph.GetMutableConsumerNodes(input_arg->Name());
+      for (auto& consumer_node : nodes) {
```

Comment:
This entire loop should be a function.

Reply:
The for loop is used once and the logic is pretty straight forward. Why do we need to put it into a function?

---
In reply to: [409795835](https://github.com/microsoft/onnxruntime/pull/3513#discussion_r409795835) [](ancestors = 409795835)